tests/tests/tier0/bluechi-restart-unit/python/monitor.py

```diff
@@ -0,0 +1,94 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+import unittest
+import subprocess
+import time
+
+from threading import Thread
+
+from dasbus.loop import EventLoop
+
+from bluechi.api import Controller, Monitor
+
+node_name_foo = "node-foo"
+
+service_simple = "simple.service"
+
+
+class TestMonitorSpecificNodeAndUnit(unittest.TestCase):
+
+    def setUp(self) -> None:
+        self.loop = EventLoop()
+        self.mgr = Controller()
+
+        self.inactive = False
+        self.active = True
+        self.startup_time = time.time()
+
+    def run_command(self, args, shell=True, **kwargs):
+        process = subprocess.Popen(
+                args,
+                shell=shell,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                **kwargs)
+        print(f"Executing of command '{process.args}' started")
+        out, err = process.communicate()
+
+        out = out.decode("utf-8")
+        err = err.decode("utf-8")
+
+        print(
+            f"Executing of command '{process.args}' finished with result '{process.returncode}',"
+            " stdout '{out}', stderr '{err}'")
+
+        return process.returncode, out, err
+
+    def process_events(self):
+        self.loop.run()
+
+    def test_monitor_specific_node_and_unit(self):
+
+        monitor_path = self.mgr.create_monitor()
+        monitor = Monitor(monitor_path=monitor_path)
+
+        def on_unit_state_changed(node: str, unit: str, active_state: str, sub_state: str, reason: str) -> None:
+            print(f"Received state change for node '{node}', unit '{unit}', "
+                  f"active_state '{active_state}', sub_state '{sub_state}'")
+            if active_state == "inactive" and self.active and not self.inactive:
+                # Active -> Inactive
+                print(f"Service '{unit}' on node '{node}' was stopped successfully")
+                self.active = False
+                self.inactive = True
+            elif active_state == "active" and self.inactive:
+                # Inactive -> Active
+                print("Service '{unit}' on node '{node}' was started successfully")
+                self.active = True
+                self.loop.quit()
+            else:
+                print(f"Ignoring status change to '{active_state}'")
+
+            # Just a failsafe to quit event loop before the whole test is timeouted to have some logs
+            current_time = time.time()
+            if (self.startup_time + 10) < current_time:
+                print(f"Loop timeout - service '{unit}' on node '{node}' was not successfully restarted on time")
+                self.loop.quit()
```

Comment:
This only gets run IF we receive some unit state changed signals, right? I think it would be better to start a separate thread which sleeps for n seconds and then quits the loop:
```python
def wait(loop: EventLoop):
    time.sleep(10)
    print(f"Loop timeout - service '{unit}' on node '{node}' was not successfully restarted on time")
    loop.quit()

# ...
t = Thread(target=self.process_events)
# mark the failsafe thread as daemon so it stops when the main process stops
failsafe_thread = threading.Thread(target=wait, args=(loop,), daemon=True)
t.start()
failsafe_thread.start()
# ...
```

Reply:
Makes sens. done